upload/upload_private_rules.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
+import argparse
+import os
+import sys
+from pathlib import Path
+from typing import Any
+from typing import Dict
+
+import jsonschema
+import requests
+import yaml
+
+SEMGREP_UPLOAD_DEPLOYMENT_ID_ENVVAR = "SEMGREP_UPLOAD_DEPLOYMENT"
+SEMGREP_UPLOAD_TOKEN_ENVVAR = "SEMGREP_TOKEN"
+SCHEMA_PATH = "rule_schema.yaml"
+SEMGREP_REGISTRY_BASE_URL = os.environ.get(
+    "SEMGREP_REGISTRY_BASE_URL", "https://semgrep.dev"
+)
+SEMGREP_REGISTRY_UPLOAD_URL = f"{SEMGREP_REGISTRY_BASE_URL}/api/registry/rule"
+SEMGREP_REGISTRY_VIEW_URL = f"{SEMGREP_REGISTRY_BASE_URL}/r/"
+
+
+def is_valid(rule: Dict[str, Any]) -> bool:
+    """
+    Returns true if the rule is valid for uploading
+
+    A valid rule definition should:
+    - only have one rule definition
+    - passes schema validation
+    """
+    if len(rule["rules"]) != 1:
+        print("Rule contains more than one rule")
+        return False
+
+    with Path(SCHEMA_PATH).open() as schema_file:
+        schema = yaml.safe_load(schema_file)
+
+    try:
+        jsonschema.validate(rule, schema, cls=jsonschema.validators.Draft7Validator)
+    except jsonschema.ValidationError as ve:
+        print(ve)
+        return False
+
+    return True
+
+
+def upload_rule(rule_file: Path, deployment_id: int, token: str) -> None:
+    """
+    Uploads rule in rule_file to private registry of deployment_id
+
+    Args:
+        rule_file: path to valid rule yaml file that contains single
+        rule to be uploaded
+        deployment_id: which deployment to upload a rule to
+        token: token with permissions to upload a rule
+    """
+    with rule_file.open("r") as file:
+        rule = yaml.safe_load(file)
+
+    if not is_valid(rule):
+        print(
+            "Rule is not valid. rule file must conform to semgrep rule schema and only contain a single rule"
+        )
+        sys.exit(1)
+
+    session = requests.Session()
+    session.headers["Authorization"] = f"Bearer {token}"
+
+    request_json = {
+        "definition": rule,
+        # below should always be defined if passed validation
+        "languages": rule["rules"][0]["languages"],
+        "deployment_id": deployment_id,
```

Comment:
Do you need a path?

Reply:
For v0 we're gonna make a simplifying assumption that there's no path and the rule name in the registry will just be orgname:rulename